@here/olp-sdk-core/lib/utils/MultiPartUploadWrapper.ts

```diff
@@ -177,9 +221,211 @@ export class MultiPartUploadWrapper {
+
+                            opts.onStatus &&
+                                opts.onStatus({
+                                    chunkId: uploadPartResponse.id,
+                                    chunkNumber,
+                                    chunkSize: buffer.byteLength,
+                                    totalChunks,
+                                    uploadedChunks
+                                });
+                        })
+                );
+            }
+
+            await Promise.all(promises);
+        }
+
+        await request.completeMultipartUpload({
+            parts,
+            url: startMultipartResponse.completeUrl,
+            multipartToken: startMultipartResponse.multipartToken,
+            layerId: opts.layerId,
+            billingTag: opts.billingTag
+        });
+
+        blobData.finally && blobData.finally();
+
         return STATUS_CODES.NO_CONTENT;
     }
+
+    private isStartMultipartResponseValid(obj: {
+        multipartToken?: string;
+        uploadPartUrl?: string;
+        completeUrl?: string;
+        statusUrl?: string;
+    }): boolean {
+        return (
+            obj.multipartToken !== undefined ||
+            (obj.uploadPartUrl !== undefined && obj.completeUrl !== undefined)
+        );
+    }
+
+    private isBlobData(input: any): boolean {
+        return (
+            typeof input.size === "function" &&
+            typeof input.readBytes === "function"
+        );
+    }
+
+    private async prepareBlobData(
+        input: string | File | Blob | ArrayBufferLike | BlobData
+    ): Promise<BlobData> {
+        if (this.isBlobData(input)) {
+            return input as BlobData;
+        }
+
+        if (typeof Buffer !== "undefined") {
+            // node.js
+            if (Buffer.isBuffer(input)) {
+                return new BufferData(input);
+            }
+
+            if (typeof input === "string") {
+                return NodeFileData.fromPath(input);
+            }
+        }
+
+        // browser
+        if (typeof input === "string") {
+            throw new Error("Unsupported input");
+        }
+
+        return input instanceof Blob
+            ? new WebData(input)
+            : new BufferData(input as ArrayBuffer);
+    }
+
+    private validateChunkSize(size = 0): number {
+        const minChunkSizeMiB = 5;
+        const mmxChunkSizeMiB = 5120;
+        const bytesInMiB = 1048576;
+        return size >= minChunkSizeMiB && size <= mmxChunkSizeMiB
+            ? size * bytesInMiB
+            : minChunkSizeMiB * bytesInMiB;
+    }
+
+    private async getUploadRequest(opts: {
+        blobVersion: number;
+        catalogHrn: string;
+        abotrSignal?: AbortSignal;
+    }): Promise<UploadRequest> {
+        const cacheKey = `request::blob::v${opts.blobVersion}::catalog::${opts.catalogHrn}`;
+        const cachedRequest = this.cachedUploadRequests.get(cacheKey);
+        if (cachedRequest) {
+            return cachedRequest;
+        }
```

Comment:
What is this storing or taking from cache?
Also please consider the common key structure: <hrn>::<...>, which means the HRN is always first.

Reply:
Cached `DatastoreRequestBuilder` instance with the configured lookup to the catalog and blob storage. Users can initialize once the wrapper and use the method `upload` for uploading different data to the different catalogs and blobs.

Reply:
If users can initialize one the wrapper, it means that you are not only storing the lookup APIs, but also the multipartToken as the multipart upload is already started. This is not good, as it does not say how long a started publication is valid and thus you might save something that is not even valid anymore. Also the lookup APIs need to be stored separately as those are valid per catalog and not per publication.

Reply:
Good catch. Will remove the cache.